### PR TITLE
cmd/snap-confine, interfaces/udev: device cgroup support for non-strict confinement

### PR DESCRIPTION
**Pull request by bboozzoo**
We previously had an implicit handling of non-strict confinement modes, such that when a devmode snap is installed, no udev devices would be tagged for it, and thus there would be no device cgroup established. With introduction of mandatory device cgroups for core24+ (and bare) bases, the --devmode confinement would no longer operate in this mode, as the snap would always end up being placed in a mandatory cgroup. Address this by introducing a `non-strict=true` attribute in /var/lib/snapd/cgroup/snap.*.device file. This is an overall improvement, as the non-strict confinement is now explicit for both old and new snaps.

### CODE REVIEW

File: cmd/snap-confine/snap-confine.c [hunk 1]
```diff
@@ -648,14 +658,21 @@ static bool is_device_cgroup_self_managed(const sc_invocation *inv)
...
 	}
 
+	sc_error *err SC_CLEANUP(sc_cleanup_error) = NULL;
 	char *self_managed_value SC_CLEANUP(sc_cleanup_string) = NULL;
-	sc_error *err = NULL;
 	if (sc_infofile_get_key
 	    (stream, "self-managed", &self_managed_value, &err) < 0) {
 		sc_die_on_error(err);
 	}
+	rewind(stream);
```
Comment (zyga) [inline cmd/snap-confine/snap-confine.c:667]: What is the point of this rewind?

**Review by zyga — APPROVED** [review]
The changes to s-c look good.

In another release we could tweak sc_infofile_ functions to call a callback instead of duplicating a string. We would then be able to drop the pointless strdup/free pairs and rewind to match.

I was wondering if we should follow suit and also drop classic=true for such snaps, so that we have no magic left in this equation.

File: interfaces/udev/backend.go [hunk 2]
```diff
@@ -153,15 +153,19 @@ func (b *Backend) Setup(appSet *interfaces.SnapAppSet, opts interfaces.Confineme
...
 
 	if udevSpec.ControlsDeviceCgroup() {
 		// The spec states that the snap can manage its own device
 		// cgroup (typically applies to container-like snaps), in which
 		// case leave a flag for snap-confine in at a known location.
-
+		deviceBuf.WriteString("# snap is allowed to manage own device cgroup.\n")
 		deviceBuf.WriteString("self-managed=true\n")
 	}
+	if (opts.DevMode || opts.Classic) && !opts.JailMode {
```
Comment (zyga) [inline interfaces/udev/backend.go:164]: Perhaps classic should be distinct from devmode.

Comment (bboozzoo) [reply interfaces/udev/backend.go:164]: This duplicates the behavior we have right here: https://github.com/snapcore/snapd/pull/13777/files/946315b92797f4bc66b0d9aed23b9f8b89481f09#diff-0aa4a229d9fd98c2fba19a47254f6e2ebf8159111468122090d63741a6598fbaR115

**Review by ernestl — APPROVED** [review]
Looks good if https://github.com/snapcore/snapd/pull/13777/files#r1548075200 is resolved.

File: cmd/snap-confine/snap-confine.c [hunk 3]
```diff
@@ -648,14 +658,21 @@ static bool is_device_cgroup_self_managed(const sc_invocation *inv)
...
+
+	char *devmode_value SC_CLEANUP(sc_cleanup_string) = NULL;
+	if (sc_infofile_get_key(stream, "devmode", &devmode_value, &err) < 0) {
+		sc_die_on_error(err);
```
Comment (alexmurray) [inline cmd/snap-confine/snap-confine.c:671]: will there be instances where the conffile exists but without this new devmode value and hence will stop the snap from launching? ie. should this just assume devmode is false if it is missing?

**Review by alexmurray — APPROVED** [review]
LGTM - one minor question around handling of existing snaps.

File: cmd/snap-confine/snap-confine.c (hunk 3 above)
Comment (bboozzoo) [reply cmd/snap-confine/snap-confine.c:671]: In the case key is missing `devmode_value` will be set to NULL, which then fails `st_streq(.., "true")` check. `sc_infofile_get_key` only returns errors if the file is malformed or when passed incorrect input arguments. We already rely on this behavior, as `devmode=true` is only added if the snap is indeed installed in devmode.

Comment (alexmurray) [reply cmd/snap-confine/snap-confine.c:671]: ah ok - thanks for clarifying that for me.

File: interfaces/udev/backend.go (hunk 2 above)
Comment (pedronis) [reply interfaces/udev/backend.go:164]: but looking there, one could argue then that the flag shouldn't be devmode=true but nonstrict=true then?

**Review by pedronis — APPROVED** [review]
thanks, slightly pedantic comment related to Zyga's comment

File: interfaces/udev/backend.go (hunk 2 above)
Comment (zyga) [reply interfaces/udev/backend.go:164]: I was thinking about the name of the flag, it's not devmode when snap is classic. I agree it should be something else than now.

Comment (bboozzoo) [reply interfaces/udev/backend.go:164]: changed to `non-strict=true`

**Pull request merged**